### ISSUE TITLE
Piper thinks changes have been made if closed in the "Select a Device" screen

### ISSUE THREAD

**Issue by gandalf3** [libratbag/piper#220]
If I start piper with two G600's plugged in (testing https://github.com/libratbag/libratbag/pull/496), then go to quit piper immediately in the "Select a Device" screen by closing the window, I get a prompt saying 

> There are unapplied changes. Are you sure you want to quit?

Clicking "Yes" causes a traceback:

```
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/gi/_propertyhelper.py", line 228, in __get__
    value = self.fget(instance)
  File "/usr/lib/python3.6/site-packages/piper/mouseperspective.py", line 71, in can_shutdown
    for profile in self._device.profiles:
AttributeError: 'NoneType' object has no attribute 'profiles'
```

Piper exits normally if I click the "Quit" button built into piper, or select a device before closing the window.

**Comment by Hjdskes** [comment]
Fixing the traceback is easy; there is no `self._device` yet because you are still in the "Select a Device" screen. However, I'd much rather fix the fact that Piper thinks there are unapplied changes in the first place.

On shutdown, Piper iterates over all its perspectives ("Screens"), and it is the mouse perspective that in turn iterates over the devices' profiles to check for the dirty bit. One of these much be set, and hence Piper displays the popup for unapplied changes. Piper gets this information from ratbagd; @phomes could it be that the (WIP) driver is always reporting unapplied changes?

**Comment by phomes** [comment]
I do not see this. There are no unapplied changes and piper quits with no warning. @gandalf3 does this only happen if there are two G600's plugged in?

**Comment by gandalf3** [comment]
@phomes I don't have any other mice at hand with which to get into the "select a mouse" screen. That said, it still happens with a single G600 if I comment [lines 66 and 67 in window.py](https://github.com/libratbag/piper/blob/master/piper/window.py#L66).

Just to be sure we're all testing the same thing:  
`ratbagctl -V`: 0.9.902 ([`1fc5d4e`](https://github.com/libratbag/libratbag/pull/446/commits/1fc5d4ecb6ca4db02a82d7d14707a6232303297d), that is, my PR for the svg rebased against latest master at [`26493ac`](https://github.com/libratbag/libratbag/commit/26493aceaf094b508cf637f3f006c01f8338c4d0))
Piper version in about dialog: 0.2.901 (3cb1d7e0393912f63697e874b884f50cc408f8fc)

**Comment by phomes** [comment]
We do use the same versions. 

I just reread this thread and see that I missed an important line. You are not exiting piper via the Quit button, but in a different way. In gnome-shell the way to trigger this is to go to the overview and closing piper by highlighting it and clicking the little X in the top right corner. We do get the warning about unapplied changes. It happens for all mice/drivers I quickly tested (G600, HID++1, steelseries).

**Comment by Hjdskes** [comment]
>  It happens for all mice/drivers I quickly tested (G600, HID++1, steelseries).

Seems to work fine for my G403. AFAIK there is no difference in closing an app through GNOME Shell's overview or the app's quit button in the top right corner.

**Comment by phomes** [comment]
exiting via gnome-shell goes through do_delete_event in window.py. Using the quit button goes trough _on_quit_button_clicked in welcomeperspective.py, which does not make us go through do_delete_event. So it is not the same path.

Maybe HID++2 does something different than the three drivers I tested.

**Comment by Hjdskes** [comment]
Thanks for the analysis @phomes. I'll look into it, but I'll be away this weekend so it's going to be next week as earliest.

**Comment by Hjdskes** [comment]
Is this issue fixed now?

**Comment by phomes** [comment]
#226 just made the exit paths consistent. Real fix for this is waiting in #227
